File: my-code/567/main.go
package main

func check(a []int, b []int) bool {
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func checkInclusion(s1 string, s2 string) bool {
	s1Length, s2Length := len(s1), len(s2)
	if s1Length > s2Length {
		return false
	}
	s2 = s2 + "a"
	target := make([]int, 26)
	index := make([]int, 26)
	for i := 0; i < s1Length; i++ {
		target[s1[i]-'a']++
		index[s2[i]-'a']++
	}
	for idx := 0; idx < s2Length-s1Length+1; idx++ {
		//fmt.Println(index)
		if check(index, target) {
			return true
		}
		index[s2[idx+s1Length]-'a'] += 1
		index[s2[idx]-'a'] -= 1
	}
	return false
}

File: my-code/567/main.py
# -*- coding:utf-8 -*-
# @Time     : 2021/05/28 18: 11
# @Author   : Ranshi
# @File     : 567. 字符串的排列.py
from collections import Counter


class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        s1_length, s2_length = len(s1), len(s2)
        s2 = s2 + "0"
        ans = Counter(s1)
        idx = Counter(s2[:s1_length])
        if s1_length > s2_length:
            return False

        for left in range(0, s2_length - s1_length + 1):
            # print(idx)
            if idx == ans:
                return True
            idx[s2[left + s1_length]] += 1
            if idx[s2[left]] == 1:
                idx.pop(s2[left])
            else:
                idx[s2[left]] -= 1
        return False


if __name__ == "__main__":
    s = Solution()
    print(s.checkInclusion("ab", "eidboaoo"))
